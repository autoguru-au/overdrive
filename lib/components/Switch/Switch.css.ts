import { style, styleVariants } from '@vanilla-extract/css';

import { focusOutlineStyle } from '../../styles/focusOutline.css';
import { overdriveTokens as vars } from '../../themes/theme.css';

const colorAccent = vars.colours.foreground.body;
const colorContrast = vars.colours.background.body;
const colorMid = vars.colours.background.neutral;
const colorLight = vars.colours.background.light;
const height = vars.space['6'];
const handleSize = '24px';

export const base = style({
	display: 'inline-block',
});

export const toggle = style([
	{
		cursor: 'pointer',
		backgroundColor: colorMid,
		height: height,
		width: `calc(2 * ${height} - 2px)`,
		borderRadius: vars.border.radius.pill,
		padding: '3px 4px',
		transition: 'background-color 0.2s cubic-bezier(0, 0, 0.2, 1) 0s',
		selectors: {
			'&:not([data-disabled]):hover': {
				backgroundColor: colorAccent,
			},
		},
	},
	focusOutlineStyle,
]);

export const toggleOn = style({
	backgroundColor: colorAccent,
});

export const disabled = style({
	backgroundColor: colorLight,
	cursor: 'not-allowed',
});

const handleScale = 'scale(0.95)';
const handleTranslate = `translateX(calc(${handleSize} - 4px))`;

export const handle = styleVariants({
	default: {
		aspectRatio: '1',
		height: '100%',
		borderRadius: vars.border.radius.full,
		backgroundColor: colorContrast,
		transition: 'transform 0.2s cubic-bezier(0, 0, 0.2, 1) 0s',
		willChange: 'transform',
		selectors: {
			[`${toggle}:not([data-disabled]):hover &`]: {
				transform: handleScale,
			},
		},
	},
	active: {
		selectors: {
			[`${toggleOn} &`]: {
				transform: handleTranslate,
			},
			[`${toggle}:not([data-disabled]):hover &`]: {
				transform: `${handleScale} ${handleTranslate}`,
			},
		},
	},
});

export const storyLabel = style({
	display: 'flex',
	gap: vars.space['2'],
});
