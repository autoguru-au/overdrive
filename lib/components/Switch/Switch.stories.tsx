import { action } from '@storybook/addon-actions';
import type { Meta, StoryObj } from '@storybook/react';

import { Switch } from '.';

const meta = {
	title: 'Primatives/Switch',
	component: Switch,
} satisfies Meta<typeof Switch>;

export default meta;

type Story = StoryObj<typeof Switch>;

const defaultArgs = {
	isDisabled: false,
	isSelected: false,
	onChange: action('onChange'),
	className: 'toggleButton-class',
};

export const Untoggled: Story = {
	args: defaultArgs,
};

export const UntoggledDisabled: Story = {
	args: {
		...defaultArgs,
		isDisabled: true,
	},
};

export const Toggled: Story = {
	args: {
		...defaultArgs,
		isSelected: true,
	},
};

export const ToggledDisabled: Story = {
	args: {
		...defaultArgs,
		isSelected: true,
		isDisabled: true,
	},
};
