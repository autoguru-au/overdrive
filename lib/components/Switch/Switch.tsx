import cx from 'clsx';
import React, { FunctionComponent, memo, useCallback, useState } from 'react';
import styles from './style.scss';

export interface IProps {
	className?: string;
	disabled?: boolean;
	toggled?: boolean;

	onChange?(value: boolean): void;
}

const SwitchComponent: FunctionComponent<IProps> = ({
	className = '',
	disabled = false,
	toggled: incomingToggled = false,
	onChange,
}) => {
	const [toggled, setToggled] = useState<boolean>(incomingToggled);

	const [prevValue, setPrevValue] = useState<boolean>(incomingToggled);

	const onToggle = useCallback(() => {
		if (disabled) {
			return;
		}
		if (typeof onChange === 'function') {
			onChange(!toggled);
		}
		setToggled(!toggled);
	}, [incomingToggled, disabled, toggled]);

	if (prevValue !== incomingToggled) {
		setToggled(incomingToggled);
		setPrevValue(incomingToggled);
	}

	return (
		<div
			className={cx(
				styles.root,
				{
					[styles.toggled]: toggled,
				},
				className
			)}
			aria-disabled={disabled}
			aria-label="toggle value"
			onClick={onToggle}>
			<span className={styles.handle} />
		</div>
	);
};

export const Switch = memo(SwitchComponent);
