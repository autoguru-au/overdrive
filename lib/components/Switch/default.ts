export { Switch as default } from './Switch';
