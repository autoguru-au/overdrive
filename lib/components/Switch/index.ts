export { Switch, IProps as IToggleButtonProps } from './Switch';
