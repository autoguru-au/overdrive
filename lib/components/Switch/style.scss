@import '../../theme/tokens/facade';

@mixin isDisabledState() {
	&[disabled],
	&[aria-disabled='true'] {
		cursor: not-allowed;

		&:not(.loading) {
			@content;
		}
	}
}

$handleSize: 24px;
$handleOffset: 2px;
$borderSize: 1px;

.root {
	font-size: var(--global--fontSize--5);
	line-height: var(--global--fontSize--5);
	background-color: getColour(gray, 100);
	border: #{$borderSize} solid getColour(gray, 300);
	border-radius: $handleSize + $handleOffset;
	width: (2 * $handleSize) + $handleOffset - 2 * ($borderSize);
	height: $handleSize + ($handleOffset * 2);
	position: relative;
	cursor: pointer;
	transition: background-color 0.2s cubic-bezier(0, 0, 0.2, 1) 0s;

	@include isDisabledState() {
		background-color: getColour(gray, 10);
		border-color: getColour(gray, 200);

		&.toggled {
			background-color: getColour(green, 200);
			border-color: getColour(green, 200);
		}

		.handle {
			border-color: getColour(gray, 200);
		}
	}

	&.toggled {
		background-color: getColour(green, 600);
		border-color: getColour(green, 700);
	}
}

.handle {
	width: $handleSize;
	height: $handleSize;
	background-color: white;
	box-shadow: var(--global--elevationBoxShadow--2);
	border: #{$borderSize} solid getColour(gray, 300);
	border-radius: 50%;
	outline: none;
	fill: white;
	padding: 0;
	position: absolute;
	left: $handleOffset - $borderSize;
	top: $handleOffset - $borderSize;
	transition: transform 0.2s cubic-bezier(0, 0, 0.2, 1) 0s;
	will-change: transform;

	.toggled & {
		transform: translateX($handleSize - (2 * $handleOffset));
	}
}
