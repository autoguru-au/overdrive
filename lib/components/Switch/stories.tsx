import { action } from '@storybook/addon-actions';
import { boolean } from '@storybook/addon-knobs';
import { storiesOf } from '@storybook/react';
import React from 'react';
import { Switch } from './Switch';

storiesOf('Components|Switch', module)
	.add('default', () => (
		<Switch
			onChange={action('onChange')}
			disabled={boolean('disabled', false)}
			toggled={boolean('toggled', false)}
		/>
	))
	.add('Un-toggled', () => <Switch toggled={false} />)
	.add('Toggled', () => <Switch toggled={true} />)
	.add('Disabled Un-toggled', () => (
		<Switch disabled={true} toggled={false} />
	))
	.add('Disabled Toggled', () => <Switch disabled={true} toggled={true} />);
