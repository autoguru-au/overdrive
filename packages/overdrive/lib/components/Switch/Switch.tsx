import clsx from 'clsx';
import * as React from 'react';
import { FunctionComponent, memo, useCallback, useState } from 'react';
import { useStyles } from 'react-treat';

import { Box, useBoxStyles } from '../Box';
import { useTextStyles } from '../Text';

import * as styleRefs from './Switch.treat';

export interface Props {
	className?: string;
	disabled?: boolean;
	toggled?: boolean;

	onChange?(value: boolean): void;
}

const SwitchComponent: FunctionComponent<Props> = ({
	className = '',
	disabled = false,
	toggled: incomingToggled = false,
	onChange,
}) => {
	const styles = useStyles(styleRefs);
	const [toggled, setToggled] = useState<boolean>(incomingToggled);

	const [prevValue, setPrevValue] = useState<boolean>(incomingToggled);

	const onToggle = useCallback(() => {
		if (disabled) {
			return;
		}

		if (typeof onChange === 'function') {
			onChange(!toggled);
		}

		setToggled(!toggled);
	}, [incomingToggled, disabled, toggled]);

	if (prevValue !== incomingToggled) {
		setToggled(incomingToggled);
		setPrevValue(incomingToggled);
	}

	return (
		<Box
			is="button"
			className={clsx(
				styles.root,
				useTextStyles({ size: '5' }),
				{
					[styles.toggled]: toggled,
					[styles.disabled.default]: disabled,
					[styles.disabled.toggled]: toggled && disabled,
				},
				className,
			)}
			borderRadius="pill"
			position="relative"
			aria-disabled={disabled}
			aria-label={`toggle ${toggled ? 'on' : 'off'}`}
			onClick={onToggle}>
			<Box
				borderWidth="1"
				position="absolute"
				borderRadius="pill"
				backgroundColour="white"
				padding="none"
				boxShadow="2"
				className={clsx(
					styles.handle.default,
					useBoxStyles({ is: 'button' }),
					useTextStyles({ colour: 'white' }),
					{
						[styles.handle.transition]: toggled,
					},
				)}
			/>
		</Box>
	);
};

export const Switch = memo(SwitchComponent);
