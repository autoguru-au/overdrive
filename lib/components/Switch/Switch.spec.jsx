import { act, fireEvent, render } from '@testing-library/react';
import * as React from 'react';
import { useState } from 'react';

import { Switch } from './Switch';

describe('<Switch />', () => {
	it('should not throw', () =>
		expect(() => render(<Switch />)).not.toThrow());

	it('should match snapshot without props', () => {
		expect(render(<Switch />).container.firstChild).toMatchSnapshot();
	});

	it('should match snapshot when un-toggled', () => {
		expect(
			render(<Switch toggled={false} />).container.firstChild,
		).toMatchSnapshot();
	});

	it('should match snapshot when un-toggled and disabled', () => {
		expect(
			render(<Switch disabled toggled={false} />).container.firstChild,
		).toMatchSnapshot();
	});

	it('should match snapshot when toggled', () => {
		expect(
			render(<Switch toggled />).container.firstChild,
		).toMatchSnapshot();
	});

	it('should match snapshot when toggled and disabled', () => {
		expect(
			render(<Switch toggled disabled />).container.firstChild,
		).toMatchSnapshot();
	});

	it.skip('should pass on className to dom element', () => {
		expect(
			render(
				<Switch className="toggleButton-class" value={10} />,
			).container.querySelector('div'),
		).toHaveClass('toggleButton-class');
	});

	it.skip('should set toggle to false by default', () => {
		expect(render(<Switch />).container.firstChild).not.toHaveClass(
			'toggled',
		);
	});

	it.skip('should be toggled on when toggled prop is set to true', () => {
		expect(render(<Switch toggled />).container.firstChild).toHaveClass(
			'toggled',
		);
	});

	it('should be enabled by default', () => {
		expect(
			render(<Switch />).container.querySelector('.disabled'),
		).not.toBeInTheDocument();
	});

	it.skip('should have aria-disabled attribute when disabled', () => {
		const { container } = render(<Switch disabled />);

		expect(container.firstChild).toHaveClass('disabled');
		expect(container.firstChild).toHaveAttribute('aria-disabled', 'true');
	});

	it('should fire change with the correct changed value when clicked', () => {
		const spyedCallback = jest.fn();

		const { container } = render(
			<Switch toggled={false} onChange={spyedCallback} />,
		);

		fireEvent.click(container.firstChild);

		expect(spyedCallback).toHaveBeenCalledWith(true);

		fireEvent.click(container.firstChild);

		expect(spyedCallback).toHaveBeenCalledWith(false);

		fireEvent.click(container.firstChild);

		expect(spyedCallback).toHaveBeenCalledWith(true);
		expect(spyedCallback).toHaveBeenCalledTimes(3);
	});

	it('should not fire change if clicked while disabled', () => {
		const spyedCallback = jest.fn();

		const { container } = render(
			<Switch disabled toggled={false} onChange={spyedCallback} />,
		);

		fireEvent.click(container.firstChild);

		expect(spyedCallback).not.toHaveBeenCalled();
	});

	it.skip('should update its value when and a value prop comes in', () => {
		const ToggleButtonWrapper = ({ setter }) => {
			const [toggled, toggledValue] = useState(false);

			setter(toggledValue);

			return <Switch toggled={toggled} />;
		};

		let setToggledValue;
		const { container } = render(
			<ToggleButtonWrapper
				setter={setter => {
					setToggledValue = setter;
				}}
			/>,
		);

		expect(container.firstChild).not.toHaveClass('toggled');

		act(() => setToggledValue(true));

		expect(container.firstChild).toHaveClass('toggled');

		act(() => setToggledValue(false));

		expect(container.firstChild).not.toHaveClass('toggled');
	});
});
