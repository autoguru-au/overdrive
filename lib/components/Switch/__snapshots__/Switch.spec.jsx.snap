// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Switch /> should match snapshot when toggled 1`] = `
<div
  aria-disabled="false"
  aria-label="toggle value"
  class="root toggled"
>
  <span
    class="handle"
  />
</div>
`;

exports[`<Switch /> should match snapshot when toggled and disabled 1`] = `
<div
  aria-disabled="true"
  aria-label="toggle value"
  class="root toggled"
>
  <span
    class="handle"
  />
</div>
`;

exports[`<Switch /> should match snapshot when un-toggled 1`] = `
<div
  aria-disabled="false"
  aria-label="toggle value"
  class="root"
>
  <span
    class="handle"
  />
</div>
`;

exports[`<Switch /> should match snapshot when un-toggled and disabled 1`] = `
<div
  aria-disabled="true"
  aria-label="toggle value"
  class="root"
>
  <span
    class="handle"
  />
</div>
`;

exports[`<Switch /> should match snapshot without props 1`] = `
<div
  aria-disabled="false"
  aria-label="toggle value"
  class="root"
>
  <span
    class="handle"
  />
</div>
`;
